### PR TITLE
Add per-category button to fill budget cells, including Goal template support

### PR DESCRIPTION
**Pull request by shall0pass**
I don't think this will pass code review as is and will need some input to get it finished.  I added a button to the category drop down "Apply Goal Template" to apply a single template to the budget.  

~~I wasn't sure exactly sure how the UX should be interpreted if multiple months are shown, so currently it will only calculate for the current month.~~
I think a button in the budget cell makes more sense.

If we can get this to work, it'll close https://github.com/actualbudget/actual/issues/958 and
https://github.com/actualbudget/actual/issues/682

This PR has changed significantly since I started it.  It now includes per category options that were previously only found in the budget summary menu.  I've added a few more, but happy to remove or modify what I have.  In light of these changes, I'm going to change the PR title.

### CODE REVIEW

File: packages/desktop-client/src/components/budget/misc.js [hunk 1]
```diff
@@ -396,6 +398,12 @@ function SidebarCategory({
                   onDelete(category.id);
                 } else if (type === 'toggleVisibility') {
                   onSave({ ...category, hidden: !category.hidden });
+                } else if (type === 'apply-single-category-template') {
+                  let month = monthUtils.currentMonth();
```
Comment (shall0pass) [inline packages/desktop-client/src/components/budget/misc.js:402]: I could use a pointer on how this should operate.  If you're on a single month view and looking at THIS month, it appears to work correctly.  If you're on a single month view and on a DIFFERENT month, it will not appear to work correctly.  If you're on a multi-month view🤷‍♂️

File: packages/loot-core/src/server/budget/app.ts [hunk 2]
```diff
@@ -28,6 +28,10 @@ app.method(
   'budget/overwrite-goal-template',
   mutator(undoable(goalActions.overwriteTemplate)),
 );
+app.method(
```
Comment (shall0pass) [inline packages/loot-core/src/server/budget/app.ts:31]: This was a convenient place to put this.  I'm not sure if it's the correct place.

File: packages/loot-core/src/server/budget/types/handlers.d.ts [hunk 3]
```diff
@@ -15,6 +15,10 @@ export interface BudgetHandlers {
 
   'budget/apply-goal-template': (...args: unknown[]) => Promise<unknown>;
 
+  'budget/apply-single-category-template': (
```
Comment (shall0pass) [inline packages/loot-core/src/server/budget/types/handlers.d.ts:18]: Again, put here for convenience.  I don't know if it's the correct place.

File: packages/desktop-client/src/components/budget/misc.js [hunk 4]
```diff
@@ -396,6 +398,12 @@ function SidebarCategory({
                   onDelete(category.id);
                 } else if (type === 'toggleVisibility') {
                   onSave({ ...category, hidden: !category.hidden });
+                } else if (type === 'apply-single-category-template') {
```
Comment (joel-jeremy) [inline packages/desktop-client/src/components/budget/misc.js:401]: Suggestion: what if we add a button/icon beside the `#template` text in the notes instead to run the template? Or add a button at the button of the notes component if it detects a `#template` text?

**Comment by shall0pass** [comment]
I was tempted to put the button in the tool-tip menu of the spent column, because then there is a clear intention as to which month you want to apply the template.  It feels like a weird place though for UX.  Also, the currently available variables available inside the code chunk are monthIndex (which seems to always be 0 or greater) and categoryId (which is different from category but I could deal with this).  I don't really know how to add new UX pieces like buttons inside other components, so I would need help doing anything like that.  This project is more of a hobby for me, so I'm learning as I go :)  

I know there's a tendency to always compare Actual to YNAB, and I try not to do it.  But in this case (thinking YNAB4 here), it has a menu list inside the budgeted cell, which feels like a natural fit for this menu option.  But again, that menu doesn't exist and I don't know how to add it.

**Comment by youngcw** [comment]
Having a menu inside the budget cell could be nice. The other auto budget options (3month average, last month) could also be added there. 
Would it be useful to add a template of X prev month average?  That way if you like to do that you have an option to do it automatically.

**Comment by shall0pass** [comment]
I haven't pushed this up yet.  I wanted some feedback.  I think the chevrons need to be invisible unless hovered, and probably centered vertically better.  I'm still looking for examples to do that.  All of the other options are hooked up and work on a per category basis.  I have this in a separate branch if you want to check it out.  https://github.com/shall0pass/actual/tree/budgetDrop

budget_per_category_options

**Comment by j-f1** [comment]
A couple of small nitpicks:

- we seem to use sentence case for button/menu item labels across the app, can you update your menu items to match?
- Adding in the button seems to pull the budget rows out of alignment with the header row:
Screenshot_2023-07-19 11 43 06

**Comment by shall0pass** [comment]
I think I have the requested changes taken care of.
image
image

I don't understand the error on the failing test.  Could you help me debug it?  It looks like sync.test.ts is the test that is failing.

**Comment by joel-jeremy** [comment]
I think the error has something to do with setScrollWidth on line 82 of `TransactionsTable.test.js`.

Small nitpick: there is a gap in the rows where e the button is located and seems the line thickness is also affected. :)

Screenshot_20230720-165113~2

**Comment by shall0pass** [comment]
I can't seem to find the right properties to turn off the border above and below the <SheetCell> component.  I think the border is being drawn twice.
image

File: packages/loot-core/src/server/budget/types/handlers.d.ts [hunk 5]
```diff
@@ -13,15 +13,13 @@ export interface BudgetHandlers {
...
 
-  'budget/apply-goal-template': (arg: {
-    month: string;
-  }) => Promise<Notification>;
+  'budget/apply-goal-template': (...args: unknown[]) => Promise<unknown>;
 
-  'budget/overwrite-goal-template': (arg: {
-    month: string;
-  }) => Promise<Notification>;
+  'budget/overwrite-goal-template': (...args: unknown[]) => Promise<unknown>;
```
Comment (j-f1) [inline packages/loot-core/src/server/budget/types/handlers.d.ts:23]: Were these changes intentional?

**Comment by j-f1** [comment]
One of the test warnings should be fixed by https://github.com/actualbudget/actual/pull/1366, but the failing tests have me stumped since you don’t seem to have changed any related code.

File: packages/loot-core/src/server/budget/types/handlers.d.ts (hunk 5 above)
Comment (shall0pass) [reply packages/loot-core/src/server/budget/types/handlers.d.ts:23]: These were giving me TS errors in the tests. I'm uninformed on how to handle these errors, but this change made them pass the test.  I added 'async' to both of those functions which caused the errors to come up.  Is there something I need to add to the function line for the test to pass?

Comment (shall0pass) [reply packages/loot-core/src/server/budget/types/handlers.d.ts]: I just reverted the changes.  I don't see the errors now.

**Comment by j-f1** [comment]
A few visual issues:

- the arrow appears to have a very light gray background. I can’t track down exactly where it’s coming from but it would be nice to avoid that if possible:
  Screenshot_2023-07-21 15 05 45
- larger monetary amounts can’t be read as easily — maybe only render the button if the cursor is hovering over the row? that way the full width would still be usable for the amount:
  Screenshot_2023-07-21 15 06 33
- the width of the input field is also constrained to include the button — maybe hide the button while editing that field?
  Screenshot_2023-07-21 15 07 34

**Comment by shall0pass** [comment]
Let me know if this feels better.

**Comment by shall0pass** [comment]
Better now :)

File: packages/loot-core/src/server/budget/actions.ts [hunk 6]
```diff
@@ -272,3 +319,9 @@ export async function setCategoryCarryover({ startMonth, category, flag }) {
     }
   });
 }
+
+export async function setSingleCategoryTemplate({ month, category }) {
+  await batchMessages(async () => {
```
Comment (shall0pass) [inline packages/loot-core/src/server/budget/actions.ts:324]: This function, when commented out, will allow the tests to pass.

**Comment by shall0pass** [comment]
The failing test is now fixed. 😅

**Comment by j-f1** [comment]
Sorry for the delay in reviewing this — it somehow keeps slipping out of my notifications. A few small issues:

- the button is still visible when typing in the field & hovering over it which would make it hard to click the insertion point around, especially if you’re typing in math.

  https://github.com/actualbudget/actual/assets/25517624/3ec411db-a5ba-4e2c-ae88-529b1b7a7a02

- moving the mouse away from the row causes the open menu to disappear — probably clicking the button open should cause it to stay visible until the menu is closed.

  https://github.com/actualbudget/actual/assets/25517624/3b3a590d-2c4f-4816-915f-b009f0a04897

**Review by j-f1 — APPROVED** [review]
LGTM — thanks for fixing so quickly!

**Comment by shall0pass** [comment]
@j-f1 Try this last one first.  @kyrias had some feedback about the arrow staying visible while the menu is open and I think it looks a bit better.

**Review by j-f1 — APPROVED** [review]
Even better!

**Comment by joel-jeremy** [comment]
Arrow doesn't seem to show on mobile

**Comment by j-f1** [comment]
I think that’s fine for now, and whoever adds edit support to the mobile budget screen can figure out how to integrate these actions.
